### PR TITLE
Data selector multi step refactor + replace field mapping with data selector

### PR DESCRIPTION
**Pull request by joneath**
First step to address issue #6183.

### CODE REVIEW

**Comment by attekei** [comment]
Ready for review in terms of both functionality and code.

**Comment by attekei** [comment]
I think that the field filter picker UI still needs some polish too.

**Review by mazameli — CHANGES_REQUESTED** [review]
Hmm, this appears to be broken to me. When I click on any field I get a JS error. Video:

[field_value_multi_step.mp4.zip](https://github.com/metabase/metabase/files/1675187/field_value_multi_step.mp4.zip)

## Cosmetic things

**The top of this popover should show the table name rather than just "fields:"**

screen shot 2018-01-29 at 12 55 07 pm

**This select looks wrong:**

screen shot 2018-01-29 at 12 43 13 pm

**The name of the database here is clickable but shouldn't be (it doesn't do anything):**

screen shot 2018-01-29 at 12 44 25 pm

File: frontend/src/metabase/redux/metadata.js [hunk 1]
```diff
@@ -480,6 +497,17 @@ export const fetchDatabasesWithMetadata = createThunkAction(FETCH_DATABASES_WITH
     };
 });
 
+const FETCH_REAL_DATABASES_WITH_METADATA = "metabase/metadata/FETCH_REAL_DATABASES_WITH_METADATA";
+export const fetchRealDatabasesWithMetadata = createThunkAction(FETCH_REAL_DATABASES_WITH_METADATA, (reload = false) => {
```
Comment (tlrobinson) [inline frontend/src/metabase/redux/metadata.js]: What does "real" databases mean?

**Review by tlrobinson — COMMENTED** [review]
Seems pretty reasonable. Just a few comments/questions

File: frontend/src/metabase/query_builder/components/DataSelector.jsx [hunk 2]
```diff
@@ -9,50 +10,46 @@ import LoadingAndErrorWrapper from "metabase/components/LoadingAndErrorWrapper"
...
+        if (props.setFieldFn) {
+            steps = [SCHEMA_STEP, TABLE_STEP, FIELD_STEP];
+        } else if (props.setSourceTableFn && props.segments) {
+            steps = [SCHEMA_AND_SEGMENTS_STEP, SEGMENT_OR_TABLE_STEP];
+        } else if (props.setSourceTableFn) {
+            steps = [SCHEMA_STEP, TABLE_STEP];
+        } else if (props.setDatabaseFn) {
+            steps = [DATABASE_STEP];
+        } else {
+            throw new Error("Can't figure out what kind of DataSelector to show")
```
[the remark this answers is not in the document]
Comment (attekei) [reply frontend/src/metabase/query_builder/components/DataSelector.jsx]: That was addressed (kind of) in my latest commits where I split each set of steps to individual React components, for instance
```
export const SchemaTableAndFieldDataSelector = (props) =>
    <DataSelector
        steps={[SCHEMA_STEP, TABLE_STEP, FIELD_STEP]}
        getTriggerElementContent={FieldTriggerContent}
        renderAsSelect={true}
        {...props}
    />
```

[the remark this answers is not in the document]
Comment (attekei) [reply frontend/src/metabase/query_builder/components/DataSelector.jsx]: But refactoring this component is an endless endeavour I think I'm going to stop here.

File: frontend/src/metabase/query_builder/components/DataSelector.jsx [hunk 3]
```diff
@@ -76,50 +75,124 @@ export default class DataSelector extends Component {
...
+
+    static defaultProps = {
+        isInitiallyOpen: false,
+        renderAsSelect: false,
+        skipDatabaseSelection: false,
+    };
+
+    componentWillMount() {
+        if (!this.props.selectedDatabaseId && this.props.databases.length === 1 && !this.props.segments) {
+            setTimeout(() => this.onChangeDatabase(0));
```
[the remark this answers is not in the document]
Comment (attekei) [reply frontend/src/metabase/query_builder/components/DataSelector.jsx]: Good call, I'll move it

File: frontend/src/metabase/redux/metadata.js (hunk 1 above)
Comment (attekei) [reply frontend/src/metabase/redux/metadata.js]: It won't include saved questions which are "virtual" databases. Will add a comment to that Redux action

**Comment by attekei** [comment]
@mazameli Fixed the issue you mentioned (alongside with some issues related to the field filter field picker). I also did the UI changes you requested. Could you do another round of reviewing?

After we've verified that the everything works as intended, I think this is ready to be merged.

**Comment by mazameli** [comment]
Not sure if this is relevant to changes made in this PR, but I'm seeing some backend syncing errors, like this one:

```
01-30 10:50:00 ERROR sync.util :: Error syncing fields for Table 6 'PUBLIC.PRODUCTS': Output of parent-id->fields does not match schema: 

           {nil #{{:special-type (not ("Valid field type" a-clojure.lang.Keyword))}}}  


("sync.sync_metadata.fields$eval47231$parent_id__GT_fields__47236.invoke(fields.clj:254)"
 "sync.sync_metadata.fields$eval47284$our_metadata__47289$fn__47290.invoke(fields.clj:280)"
 "sync.sync_metadata.fields$eval47284$our_metadata__47289.invoke(fields.clj:274)"
 "sync.sync_metadata.fields$eval47350$sync_fields_for_table_BANG___47359$fn__47362$fn__47363.invoke(fields.clj:309)"
 "sync.util$do_with_error_handling.invokeStatic(util.clj:124)"
 "sync.util$do_with_error_handling.invoke(util.clj:119)"
 "sync.sync_metadata.fields$eval47350$sync_fields_for_table_BANG___47359$fn__47362.invoke(fields.clj:306)"
 "sync.sync_metadata.fields$eval47350$sync_fields_for_table_BANG___47359.invoke(fields.clj:301)"
 "sync.sync_metadata.fields$eval47388$sync_fields_BANG___47393$fn__47394.invoke(fields.clj:318)"
 "sync.sync_metadata.fields$eval47388$sync_fields_BANG___47393.invoke(fields.clj:314)"
 "sync.sync_metadata$eval47989$sync_db_metadata_BANG___47994$fn__47995$fn__47996.invoke(sync_metadata.clj:28)"
 "sync.util$do_with_error_handling.invokeStatic(util.clj:124)"
 "sync.util$do_with_error_handling.invoke(util.clj:119)"
 "sync.util$do_with_error_handling.invokeStatic(util.clj:122)"
 "sync.util$do_with_error_handling.invoke(util.clj:119)"
 "driver$fn__29278.invokeStatic(driver.clj:222)"
 "driver$fn__29278.invoke(driver.clj:222)"
 "driver$eval29058$fn__29197$G__29023__29206.invoke(driver.clj:47)"
 "sync.util$sync_in_context$fn__45041.invoke(util.clj:115)"
 "sync.util$with_db_logging_disabled$fn__45038.invoke(util.clj:106)"
 "sync.util$with_start_and_finish_logging$fn__45033.invoke(util.clj:92)"
 "sync.util$with_sync_events$fn__45030.invoke(util.clj:75)"
 "sync.util$with_duplicate_ops_prevented$fn__45021.invoke(util.clj:54)"
 "sync.util$do_sync_operation.invokeStatic(util.clj:142)"
 "sync.util$do_sync_operation.invoke(util.clj:139)"
 "sync.sync_metadata$eval47989$sync_db_metadata_BANG___47994$fn__47995.invoke(sync_metadata.clj:23)"
 "sync.sync_metadata$eval47989$sync_db_metadata_BANG___47994.invoke(sync_metadata.clj:20)"
 "task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:36)")
```

**Comment by mazameli** [comment]
A few more notes:

**Tables here should have right-pointing arrows to show there's a subsequent step. See the way we do schema selection for reference:**

screen shot 2018-01-30 at 11 19 26 am

**Fields here shouldn't have table icons; they should have their corresponding field type icon:**

screen shot 2018-01-30 at 11 20 54 am

**The selected field here should show the friendly display name for the field, not its underlying name in the DB:**

screen shot 2018-01-30 at 11 22 04 am

**Comment by mazameli** [comment]
Functionality seems fixed now. Remaining requested changes are all cosmetic.

**Comment by attekei** [comment]
@maz I fixed all the cosmetic issues! About your backend-related comment:

> Not sure if this is relevant to changes made in this PR, but I'm seeing some backend syncing errors, like this one:

This PR only has frontend changes so I don't know what might have caused that 😕

**Comment by attekei** [comment]
I prepared for merge by squashing the old commits to more useful ones. 

And thanks to @joneath for the great job you did with refactoring `DataSelector`, it was easy to continue from where you were left!

**Review by salsakran — APPROVED** [review]
Metadata loading is FAR more sane with this! Thanks @joneath + @attekei

**Review by mazameli — APPROVED** [review]

**Comment by mazameli** [comment]
Looks good. I'll merge it. I am seeing some weirdness with how the field filter selector's height causes the page to scroll, but I don't believe it's a regression so we'll look at that in a separate issue.

**Comment by attekei** [comment]
> I am seeing some weirdness with how the field filter selector's height causes the page to scroll, but I don't believe it's a regression so we'll look at that in a separate issue.

As far as I know, it's caused by the combination of a sidebar with scrolling content and long lists. Similar scroll issues aren't present elsewhere where the same data picker component is used.

**Pull request merged**